[DYOD] Add generic implementation of Parallel Inplace Merge Sort

niklasmohrin opened the pull request:
These changes are extracted into a separate PR from our project work on the window function operator, as discussed with supervisors. The main PR will be opened later :)

I broke down the changes into three clean commits, so that you can keep them if you want, but feel free to squash anyways if you feel like it.

Since this uses `span` (and other C++20 features) again, gcc-9 doesn't work. Should I also include the commit updating it here, or open another PR, or do something completely different?

niklasmohrin committed:
Replace gcc 9 with gcc 10

We want to use features that gcc 9 doesn't support. Commands to
reproduce:

```sh
sed -i 's/gcc-9/gcc-10/g' Jenkinsfile Dockerfile install_dependencies.sh
sed -i 's/g++-9/g++-10/g' Jenkinsfile Dockerfile install_dependencies.sh
sed -i 's/gcc9/gcc10/g' Jenkinsfile Dockerfile install_dependencies.sh
sed -i 's/g++9/g++10/g' Jenkinsfile Dockerfile install_dependencies.sh
```

niklasmohrin committed:
Add `BooleanComparator` and `ThreeWayComparator` concepts

niklasmohrin committed:
Add `SmallMinHeap`

niklasmohrin committed:
Add `parallel_inplace_merge_sort`

Bouncner commented:
Given this list (https://en.cppreference.com/w/cpp/compiler_support), GCC 10 should at least support `std::span`. Do you know if we need more than GCC 10?

niklasmohrin commented:
In our intermediate review PR, we got it to get past the `span` error, but then it gcc-ed on us again with the "control reaches end of non-void function"; I can include the gcc-10 commit here and we can see what happens? (if it works, I can also open another PR and rebase this one)

niklasmohrin committed:
Fix missing includes

niklasmohrin committed:
Fix most linter complaints

niklasmohrin committed:
Replace deterministic test with another one that doesn't `#include <random>`

niklasmohrin commented:
Okay, works with gcc-10, but somehow clang-11 gives the error `member reference base type 'std::span' is not a structure or union` when trying to use `std::span(something).subspan(anotherthing)`. Not sure what is going on, the same line seems to work on its own on https://godbolt.org/z/nxnzabxcr :thinking: 

The full file starts working with clang 13 and breaks for different reasons in clang 11 and 12 on godbolt: https://godbolt.org/z/qdvq61WWs

niklasmohrin committed:
Try clang 13

Command to reproduce:
```bash
sed -i -e 's/clang++-11/clang++-13/g' -e 's/clang-11/clang-13/g' -e 's/clang++11/clang++13/g' -e 's/clang11/clang13/g' Jenkinsfile Dockerfile install_dependencies.sh
```

niklasmohrin committed:
Another missing import

niklasmohrin committed:
empty commit

niklasmohrin committed:
empty commit

niklasmohrin committed:
Update `DEPENDENCIES.md`

niklasmohrin committed:
Update more occurrences of clang 11

niklasmohrin committed:
empty commit

niklasmohrin committed:
Remove coverage step for now (clang 13 coverage broken, clang 11 compilation broken)

finnschoellkopf committed:
Restart CI

tjeyy commented:
It looks like nobody is working on this PR right now.
